onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorSize.cpp

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "precomp.h"
+
+namespace Dml
+{
+
+class DmlOperatorSize : public DmlOperator, SizeHelper
+{
+public:
+    DmlOperatorSize(const MLOperatorKernelCreationContext& kernelCreationContext)
+    :   DmlOperator(kernelCreationContext),
+        SizeHelper(kernelCreationContext, kernelCreationContext.GetTensorShapeDescription())
+    {
+        ML_CHECK_VALID_ARGUMENT(kernelCreationContext.GetInputCount() == 1, "Size expects 1 input tensor.");
+        ML_CHECK_VALID_ARGUMENT(kernelCreationContext.GetOutputCount() == 1, "Size expects 1 output tensor.");
+        DmlOperator::Initialize(kernelCreationContext);
+    }
+
+    // Takes a tensor as input and outputs a scalar int64 tensor containing the size of the input tensor.
+    void Compute(const MLOperatorKernelContext& kernelContext)
+    {
+        std::vector<IMLOperatorTensor*> inputTensors = GetInputTensors(kernelContext);
+        std::vector<IMLOperatorTensor*> outputTensors = GetOutputTensors(kernelContext);
+        assert(inputTensors.size() == 1);
+        assert(outputTensors.size() == 1);
+
+        // Get input shape as a vector and compute the number of elements
+        const IMLOperatorTensor* inputTensor = inputTensors[0];
+        const uint32_t inputDimCount = inputTensor->GetDimensionCount();
+        std::vector<uint32_t> inputShape(inputDimCount);
+        THROW_IF_FAILED(inputTensor->GetShape(inputDimCount, inputShape.data()));
+        uint32_t numElements = std::accumulate(inputShape.begin(), inputShape.end(), 1, std::multiplies<uint32_t>());
+
+        // Write the number of elements to output data
+        IMLOperatorTensor* outputTensor = outputTensors[0];
+        int64_t* outputData = reinterpret_cast<int64_t*>(outputTensor->GetData());
+        outputData[0] = (int64_t)numElements;
```

Comment:
A sanity check here would be nice before writing directly to memory without checking the EP or byte size, like `ReadScalarTensorData` checking `IsCpuData` and the byte size. (it should be correct by the time it reaches here 🤞, but distant sections of code change over time).